test/unit/BuilderTest.php

```diff
@@ -110,6 +110,31 @@ public function permittedForCanReplicateItemOnHeader()
         self::assertEquals(['aud' => new EqualsTo('aud', 'test')], $token->getClaims());
     }
 
+    /**
+     * @test
+     *
+     * @covers ::__construct
+     * @covers ::permittedFor
+     * @covers ::setRegisteredClaim
+     * @covers ::configureClaim
+     * @covers ::createSignature
+     * @covers ::convertItems
+     *
+     * @uses \Lcobucci\JWT\Builder::getToken
+     */
+    public function permittedForWillExplainDeprecationAndCastArrayToString()
+    {
+        $this->expectDeprecation('Using arrays for audience claims is not really supported in 3.4 and will behave different in 4.0.');
+
+        $builder = $this->createBuilder();
+        $builder->permittedFor(['test1', 'test2']);
+
+        $token = $builder->getToken(new None(), Key\InMemory::plainText(''));
+
+        self::assertEquals(['typ' => 'JWT', 'alg' => 'none'], $token->getHeaders());
+        self::assertEquals(['aud' => new EqualsTo('aud', 'Array')], $token->getClaims());
```

Comment:
If I recall correctly, as soon as the deprecation occurs, the test halts, so later code can be deleted.
Moreover I suggest to expect the deprecation right before the call that raises it, so previous code cannot fall into a false positive case:
```suggestion
        $builder = $this->createBuilder();

        $this->expectDeprecation('Using arrays for audience claims is not really supported in 3.4 and will behave different in 4.0.');
        $builder->permittedFor(['test1', 'test2']);
```

Reply:
Basically I copied an existing test case right from above, which has the exact same structure that you see in the new test case.

In fact the test execution continues, so tests further down the code are still being used. This is not the same as expecting exceptions, the deprecation code will actually collect more than one and compare if they are all correctly expected.

One could argue that testing the headers has nothing to do with the test case, but this will only remove one line.

Reply:
Ok then